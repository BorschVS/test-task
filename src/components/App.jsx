import { lazy } from 'react';
import { Route, Routes } from 'react-router-dom';

import { Layout } from './Layout';

const Home = lazy(() => import('../pages/Home'));
const Flights = lazy(() => import('../pages/Flights'));
const Hotels = lazy(() => import('../pages/Hotels'));

const App = () => (
  <Routes>
    <Route path="/" element={<Layout />}>
      <Route index element={<Home />} />
      <Route path="flights" element={<Flights />} />
      <Route path="hotels" element={<Hotels />} />
    </Route>
  </Routes>
);

export default App;
