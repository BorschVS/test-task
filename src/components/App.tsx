import { FC, lazy } from 'react';
import { Route, Routes } from 'react-router-dom';

import { Layout } from 'components';

import { Page } from 'types/interfaces';

const Home = lazy(() => import('../pages/Home'));
const Flights = lazy(() => import('../pages/Flights'));
const Hotels = lazy(() => import('../pages/Hotels'));

const pages: Page[] = [
  { path: '/', element: <Home />, index: true },
  { path: 'flights', element: <Flights /> },
  { path: 'hotels', element: <Hotels /> },
];

const App: FC = () => (
  <Routes>
    <Route path="/" element={<Layout />}>
      {pages.map((page) =>
        page.index ? (
          <Route key={page.path} index element={page.element} />
        ) : (
          <Route key={page.path} path={page.path} element={page.element} />
        )
      )}
    </Route>
  </Routes>
);

export default App;
