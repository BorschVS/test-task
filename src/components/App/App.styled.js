import styled from "@emotion/styled";

export const MainBox = styled.div`
  display: flex;
  justify-content: center;
  gap: 32px;
  align-items: start;
`;

export const GroupBox = styled.div`
  display: flex;
  flex-direction: column;
  gap: 24px;
`;
