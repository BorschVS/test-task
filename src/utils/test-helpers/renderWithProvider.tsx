import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ReactNode } from 'react';
import store from 'redux/configureStore';

export const renderWithProvider = (
  component: ReactNode
): RenderResult & { asFragment: () => DocumentFragment } => {
  const utils = render(<Provider store={store}>{component}</Provider>);

  return {
    ...utils,
    asFragment: utils.asFragment,
  };
};
