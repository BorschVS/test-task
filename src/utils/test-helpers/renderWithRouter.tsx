import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import store from 'redux/configureStore';
import { ReactNode } from 'react';
import { App } from 'components';

export const renderWithRouter = (
  component: ReactNode,
  initialRoute: string = '/'
): RenderResult & { asFragment: () => DocumentFragment } => {
  const utils = render(
    <Provider store={store}>
      <MemoryRouter initialEntries={[initialRoute]}>
        <App />
        {component}
      </MemoryRouter>
    </Provider>
  );

  return {
    ...utils,
    asFragment: utils.asFragment,
  };
};
