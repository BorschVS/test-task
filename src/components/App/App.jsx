import { Container, Avatar, Box } from "@mui/material";
import TransferFilter from "../TransferFiler/TransferFilter";
import PlaneImg from "../../images/plane.png";
import CategoryFilter from "../CategoryFilter/CategoryFilter";
import FlightCard from "../FlightCard/FlightCard";
import { GroupBox, MainBox } from "./App.styled";

const App = () => {
  return (
    <Container maxWidth="md">
      <Avatar
        src={PlaneImg}
        alt="The company logo"
        sx={{ display: "block", margin: "40px auto" }}
      />
      <MainBox>
        <TransferFilter />
        <GroupBox display="flex" flexDirection="column" gap={3}>
          <CategoryFilter />
          <FlightCard />
          <FlightCard />
          <FlightCard />
        </GroupBox>
      </MainBox>
    </Container>
  );
};

export default App;
