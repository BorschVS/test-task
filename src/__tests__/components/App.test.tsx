import '@testing-library/jest-dom';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithRouter } from 'utils/test-helpers/renderWithRouter';

import App from 'components/App';

describe('Router page rendering test', () => {
  test('Should determine home page', async () => {
    renderWithRouter(<App />, '/');

    const homePages = await screen.findAllByTestId('home-page');
    homePages.forEach((page) => expect(page).toBeInTheDocument());
  });

  test('Should open Flights page when clicking on button "FLY"', async () => {
    renderWithRouter(<App />, '/');

    const flightsLinks = screen.getAllByTestId('flights-link');
    flightsLinks.forEach((link) => {
      if (link.textContent === 'FLY!') {
        userEvent.click(link);
      }
    });

    const flightsPages = await screen.findAllByTestId('flights-page');

    await waitFor(() =>
      flightsPages.forEach((page) => expect(page).toBeInTheDocument())
    );
  });

  test('Should determine flights page', async () => {
    renderWithRouter(<App />, '/flights');

    const flightsPages = await screen.findAllByTestId('flights-page');
    await waitFor(() =>
      flightsPages.forEach((page) => expect(page).toBeInTheDocument())
    );
  });

  test('Should determine hotels page', async () => {
    renderWithRouter(<App />, '/hotels');

    const hotelsPages = await screen.findAllByTestId('hotels-page');
    await waitFor(() =>
      hotelsPages.forEach((page) => expect(page).toBeInTheDocument())
    );
  });

  test('Should match App snapshot', () => {
    const { asFragment } = renderWithRouter(<App />);

    expect(asFragment()).toMatchSnapshot();
  });
});
