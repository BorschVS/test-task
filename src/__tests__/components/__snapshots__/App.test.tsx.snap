// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router page rendering test Should match App snapshot 1`] = `
<DocumentFragment>
  <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorInfo MuiAppBar-positionStatic css-1wwtsp3-MuiPaper-root-MuiAppBar-root"
      data-testid="responsive-app-bar"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
        >
          <a
            data-testid="home-link"
            href="/"
          >
            <span>
              <div
                class="MuiAvatar-root MuiAvatar-circular css-gh7mkj-MuiAvatar-root"
              >
                <img
                  alt="Lets fly logo"
                  class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                  src="/images/plane.png"
                />
              </div>
            </span>
          </a>
          <div
            class="MuiBox-root css-15ytkbm"
          >
            <a
              data-testid="flights-link"
              href="/flights"
              style="text-decoration: none;"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1q39md6-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Flights
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </a>
            <a
              data-testid="hotels-link"
              href="/hotels"
              style="text-decoration: none;"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1q39md6-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Hotels
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </a>
          </div>
          <div
            class="MuiBox-root css-iqpydz"
          >
            <button
              aria-controls="menu-appbar"
              aria-haspopup="true"
              aria-label="account of current user"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </header>
    <main
      data-testid="home-page"
    >
      <div
        class="MuiBox-root css-1duo1bq"
      >
        <h1
          class="MuiTypography-root MuiTypography-h1 css-1chbnjg-MuiTypography-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 css-jlad6d-MuiTypography-root"
          >
            LetsFly!
          </span>
           Airlines
        </h1>
        <h2
          class="MuiTypography-root MuiTypography-h2 css-c0cjz3-MuiTypography-root"
        >
          PRESS THE BUTTON TO FLY!
        </h2>
        <a
          data-testid="flights-link"
          href="/flights"
          style="text-decoration: none;"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-fbc4mu-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            FLY!
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
    </main>
  </div>
  <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorInfo MuiAppBar-positionStatic css-1wwtsp3-MuiPaper-root-MuiAppBar-root"
      data-testid="responsive-app-bar"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
        >
          <a
            data-testid="home-link"
            href="/"
          >
            <span>
              <div
                class="MuiAvatar-root MuiAvatar-circular css-gh7mkj-MuiAvatar-root"
              >
                <img
                  alt="Lets fly logo"
                  class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                  src="/images/plane.png"
                />
              </div>
            </span>
          </a>
          <div
            class="MuiBox-root css-15ytkbm"
          >
            <a
              data-testid="flights-link"
              href="/flights"
              style="text-decoration: none;"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1q39md6-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Flights
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </a>
            <a
              data-testid="hotels-link"
              href="/hotels"
              style="text-decoration: none;"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1q39md6-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Hotels
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </a>
          </div>
          <div
            class="MuiBox-root css-iqpydz"
          >
            <button
              aria-controls="menu-appbar"
              aria-haspopup="true"
              aria-label="account of current user"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </header>
    <main
      data-testid="home-page"
    >
      <div
        class="MuiBox-root css-1duo1bq"
      >
        <h1
          class="MuiTypography-root MuiTypography-h1 css-1chbnjg-MuiTypography-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 css-jlad6d-MuiTypography-root"
          >
            LetsFly!
          </span>
           Airlines
        </h1>
        <h2
          class="MuiTypography-root MuiTypography-h2 css-c0cjz3-MuiTypography-root"
        >
          PRESS THE BUTTON TO FLY!
        </h2>
        <a
          data-testid="flights-link"
          href="/flights"
          style="text-decoration: none;"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-fbc4mu-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            FLY!
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
